src/main/java/io/reactivex/internal/operators/single/SingleTimeout.java

```diff
@@ -43,97 +44,116 @@ public SingleTimeout(SingleSource<T> source, long timeout, TimeUnit unit, Schedu
+                this.actual = actual;
             }
 
             @Override
             public void onSubscribe(Disposable d) {
-                set.add(d);
+                DisposableHelper.setOnce(this, d);
             }
 
             @Override
-            public void onSuccess(T value) {
-                set.dispose();
-                s.onSuccess(value);
+            public void onSuccess(T t) {
+                actual.onSuccess(t);
             }
 
+            @Override
+            public void onError(Throwable e) {
+                actual.onError(e);
+            }
         }
-    }
 
-    final class TimeoutObserver implements SingleObserver<T> {
+        TimeoutMainObserver(SingleObserver<? super T> actual, SingleSource<? extends T> other) {
+            this.actual = actual;
+            this.other = other;
+            this.task = new AtomicReference<Disposable>();
+            if (other != null) {
+                this.fallback = new TimeoutFallbackObserver<T>(actual);
+            } else {
+                this.fallback = null;
+            }
+        }
 
-        private final AtomicBoolean once;
-        private final CompositeDisposable set;
-        private final SingleObserver<? super T> s;
+        @Override
+        public void run() {
+            Disposable d = get();
+            if (d != DisposableHelper.DISPOSED && compareAndSet(d, DisposableHelper.DISPOSED)) {
+                if (d != null) {
+                    d.dispose();
+                }
+                SingleSource<? extends T> other = this.other;
+                if (other == null) {
+                    actual.onError(new TimeoutException());
+                } else {
+                    this.other = null;
+                    other.subscribe(fallback);
+                }
+            }
+        }
 
-        TimeoutObserver(AtomicBoolean once, CompositeDisposable set, SingleObserver<? super T> s) {
-            this.once = once;
-            this.set = set;
-            this.s = s;
+        @Override
+        public void onSubscribe(Disposable d) {
+            DisposableHelper.setOnce(this, d);
         }
 
         @Override
-        public void onError(Throwable e) {
-            if (once.compareAndSet(false, true)) {
-                set.dispose();
-                s.onError(e);
+        public void onSuccess(T t) {
+            Disposable d = get();
+            if (d != DisposableHelper.DISPOSED && compareAndSet(d, DisposableHelper.DISPOSED)) {
+                DisposableHelper.dispose(task);
+                actual.onSuccess(t);
             }
         }
 
         @Override
-        public void onSubscribe(Disposable d) {
-            set.add(d);
+        public void onError(Throwable e) {
+            Disposable d = get();
+            if (d != DisposableHelper.DISPOSED && compareAndSet(d, DisposableHelper.DISPOSED)) {
+                DisposableHelper.dispose(task);
+                actual.onError(e);
+            }
```

Comment:
Shouldn't we deliver error to `RxJavaPlugins.onError()` in `else` block?

Reply:
Indeed. Fixed.